staticaddr/deposit/fsm.go

```diff
@@ -0,0 +1,297 @@
+package deposit
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/btcsuite/btcd/txscript"
+	"github.com/btcsuite/btcd/wire"
+	"github.com/lightninglabs/lndclient"
+	"github.com/lightninglabs/loop/fsm"
+	"github.com/lightninglabs/loop/staticaddr"
+	"github.com/lightninglabs/loop/staticaddr/address"
+	"github.com/lightninglabs/loop/staticaddr/script"
+	"github.com/lightningnetwork/lnd/input"
+	"github.com/lightningnetwork/lnd/keychain"
+)
+
+const (
+	DefaultObserverSize = 20
+)
+
+var (
+	ErrProtocolVersionNotSupported = errors.New("protocol version not " +
+		"supported")
+)
+
+// States.
+var (
+	Deposited = fsm.StateType("Deposited")
+
+	PublishExpiredDeposit = fsm.StateType("PublishExpiredDeposit")
+
+	WaitForExpirySweep = fsm.StateType("WaitForExpirySweep")
+
+	Expired = fsm.StateType("Expired")
+
+	Failed = fsm.StateType("DepositFailed")
+)
+
+// Events.
+var (
+	OnStart           = fsm.EventType("OnStart")
+	OnExpiry          = fsm.EventType("OnExpiry")
+	OnExpiryPublished = fsm.EventType("OnExpiryPublished")
+	OnExpirySwept     = fsm.EventType("OnExpirySwept")
+	OnRecover         = fsm.EventType("OnRecover")
+)
+
+// FSM is the state machine that handles the instant out.
+type FSM struct {
+	*fsm.StateMachine
+
+	cfg *ManagerConfig
+
+	deposit *Deposit
+
+	params *address.Parameters
+
+	address *script.StaticAddress
+
+	ctx context.Context
+
+	blockNtfnChan chan uint32
+
+	finalizedDepositChan chan wire.OutPoint
+}
+
+// NewFSM creates a new state machine that can action on all static address
+// feature requests.
+func NewFSM(ctx context.Context, deposit *Deposit, cfg *ManagerConfig,
```

Comment:
So IIUC the plan is to have an FSM per deposit, meaning we'd sweep them back individually if they expire? Perhaps worth thinking about efficiency here, so we could possibly have a single FSM and sweep back all expired deposits? wdyt? cc: @sputn1ck

Reply:
Yes your understanding is correct. Since each deposit has the same relative timelock but most likely a different effective timelock we'd somehow have to batch all the expired deposits. So the user would have to wait an additional amount of blocks until our threshold for sweeping timouts is reached and a bunch of deposits is batched together.